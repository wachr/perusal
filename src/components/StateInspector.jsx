import { h } from "preact";
import { useState } from "preact/hooks";

import ExpandLessIcon from "@mui/icons-material/ExpandLess";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import Paper from "@mui/material/Paper";
import TextField from "@mui/material/TextField";
import PropTypes from "prop-types";

import FileStorage from "./FileStorage";

const StateInspector = ({ nodeState, setNode }) => {
  const [collapsed, setCollapsed] = useState(false);
  const stateString = JSON.stringify(nodeState, null, "\t");
  return (
    <div data-testid="StateInspector-div">
      <Box sx={{ mx: "5%" }}>
        <Paper elevation={0}>
          <Box sx={{ margin: "5pt" }}>
            <Box
              sx={{
                display: "flex",
                flexDirection: "row-reverse",
                marginRight: "-5pt",
              }}>
              <Button
                onClick={() => setCollapsed(!collapsed)}
                aria-label="show or hide state">
                {collapsed ? <ExpandMoreIcon /> : <ExpandLessIcon />}
              </Button>
              <FileStorage nodeState={nodeState} setNode={setNode} />
            </Box>
            {!collapsed && (
              <TextField
                multiline
                margin="normal"
                variant="outlined"
                label="app state"
                aria-label="app state"
                fullWidth
                minRows={4}
                maxRows={20}
                defaultValue={stateString}
                inputProps={{ readOnly: true }}
              />
            )}
          </Box>
        </Paper>
      </Box>
    </div>
  );
};

StateInspector.propTypes = {
  nodeState: PropTypes.oneOfType([
    PropTypes.shape(),
    PropTypes.array,
    PropTypes.string,
  ]).isRequired,
  setNode: PropTypes.func.isRequired,
};

export default StateInspector;
