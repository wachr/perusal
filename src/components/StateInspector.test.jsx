import { h } from "preact";
import { render, screen } from "@testing-library/preact";

import StateInspector from "./StateInspector";

describe("StateInspector", () => {
  it("should render properly", async () => {
    const state = {
      title: "title",
      content: "content",
      subtopics: {
        "subtopic A": {
          title: "subtopic A",
          content: "subtopic details"
        },
        "subtopic B": {
          title: "subtopic B"
        }
      }
    };
    const mockSetter = jest.fn();
    render(<StateInspector nodeState={state} setNode={mockSetter} />);

    const subject = await screen.getByTestId("StateInspector-div");
    expect(subject).toMatchSnapshot();
  });
});
