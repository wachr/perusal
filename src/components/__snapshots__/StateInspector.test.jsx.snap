// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateInspector should render properly 1`] = `
<div
  data-testid="StateInspector-div"
>
  <div
    class="MuiBox-root css-k0qjc5"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-ee20y-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-109982z"
      >
        <div
          class="MuiBox-root css-d3kwzx"
        >
          <button
            aria-label="show or hide state"
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandLessIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            data-testid="FileStorage-div"
          >
            <div
              class="MuiButtonGroup-root MuiButtonGroup-outlined css-uxh582-MuiButtonGroup-root"
              role="group"
            >
              <button
                class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButtonBase-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="FileUploadIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z"
                  />
                </svg>
                <span
                  class="MuiTypography-root MuiTypography-button css-1f0on15-MuiTypography-root"
                >
                  Import State
                </span>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButtonBase-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="FileDownloadIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"
                  />
                </svg>
                <span
                  class="MuiTypography-root MuiTypography-button css-1f0on15-MuiTypography-root"
                >
                  Export State
                </span>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          aria-label="app state"
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-1"
            id="mui-1-label"
          >
            app state
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <textarea
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-1"
              readonly=""
              style="height: 0px; overflow: hidden;"
            />
            <textarea
              aria-hidden="true"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
              tabindex="-1"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1z7n62"
              >
                <span>
                  app state
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
